rubix-spi/src/main/java/com/qubole/rubix/spi/fop/ObjectPoolPartition.java

```diff
@@ -58,40 +61,120 @@ public ObjectPoolPartition(ObjectPool<T> pool, PoolConfig config,
     }
   }
 
-  public BlockingQueue<Poolable<T>> getObjectQueue()
+  public void returnObject(Poolable<T> object)
   {
-    return objectQueue;
+    if (!objectFactory.validate(object.getObject())) {
+      log.debug(String.format("Invalid object for host %s removing %s ", object.getHost(), object));
+      decreaseObject(object);
+      // Compensate for the removed object. Needed to prevent endless wait when in parallel a borrowObject is called
+      increaseObjects(1, false);
+      return;
+    }
+
+    log.debug(String.format("Returning object %s to queue of host %s. Queue size: %d", object, object.getHost(), objectQueue.size()));
+    try {
+      if (!objectQueue.offer(object, 5, TimeUnit.SECONDS)) {
```

Comment:
we can make the timeout configurable in PoolConfig.

Reply:
i avoided doing that, this is only during the return path and we have a fallback to destroy object if we ever hit this case. Additional config will never be tweaked by users.

Reply:
This timeout should be < 5s I think - since it is not the same case as borrowing. I think there can be two cases where this can take time, either the number of objects created are > max size or the thread is unable to acquire the lock.
For the first case, we should instantly destroy the object and for the second one the timeout could be less.

Reply:
The second case shouldn't ever happen. I agree this could be very aggressive, as aggressive as try-and-return.  I will change this to `offer(E e)` call

Reply:
done